userland/libtock/ble.c

```diff
@@ -1,43 +1,148 @@
+/*
+ * BLE setup functions
+ */
+
 #include "ble.h"
+#include "tock.h"
+#include <stdbool.h>
+#include <stdint.h>
 #include <stdio.h>
 #include <string.h>
 
-int ble_adv_set_txpower(BLE_TX_Power_t power) {
-  return command(DRIVER_RADIO, BLE_CFG_TX_POWER, power);
+#define MAX_SIZE 31
+
+
+/*******************************************************************************
+ *   INTERNAL BLE HELPER FUNCTION Prototypes
+ *
+ *   s_   - static (file scope)
+ ******************************************************************************/
+static int s_ble_cfg_addr (uint8_t *addr, uint8_t addr_len);
+static int s_ble_cfg_flags (uint8_t flags);
+static int s_ble_cfg_adv_itv (uint16_t advertising_itv_ms);
+static int s_ble_cfg_gap_data (GapAdvertisementData_t header, uint8_t *data, uint8_t len);
+
+
+/*******************************************************************************
+ *   USER-SPACE BLE API
+ ******************************************************************************/
+
+int ble_initialize(uint16_t advertising_itv_ms, bool discoverable, uint8_t *addr,
+                   uint8_t addr_len) {
+  int err;
+
+  // configure advertisement interval
+  // if the interval is less than 20 or bigger than 10240 to kernel
+  // will use default value
+  err = s_ble_cfg_adv_itv(advertising_itv_ms);
+  if (err < TOCK_SUCCESS) return err;
+
+  uint8_t flags = BREDR_NOT_SUPPORTED;
+
+  // configure advertisement flags in the packet
+  if (discoverable) {
+    flags |= LE_GENERAL_DISCOVERABLE;
+  }
+
+  err = s_ble_cfg_flags(flags);
+  if (err < TOCK_SUCCESS) return err;
+
+  // if this is NULL advertisment address {0, 0, 0, 0, 0, 0}  will be used
```

Comment:
Does `{0, 0, 0, 0, 0, 0}` have a particular meaning? If so, processes should be explicit rather than using `NULL`. If not, than maybe this should promise less---i.e. it will use some preconfigured address either from another process or from the kernel.

Reply:
No, it means that kernel will use preconfigured adress which is better.

I will change.

Reply:
I didn't really think about it before, but it's probably a good idea for the kernel to set the BLE address of the device and for the application to have no input on it.

However, this isn't how the nRF SDK works, so if we made that change, it would be a difference between serialized and native applications.